zeppelin-web/src/app/visualization/visualization.js

```diff
@@ -12,161 +12,161 @@
+   * Don't need to destroy this.targetEl.
+   */
+  destroy() {
+    // override this
+  };
+
+  /**
+   * return {
+   *   template : angular template string or url (url should end with .html),
+   *   scope : an object to bind to template scope
+   * }
+   */
+  getSetting() {
+    // override this
+  };
+
+  /**
+   * Activate. invoked when visualization is selected
+   */
+  activate() {
+    if (!this._active || this._dirty) {
+      this.refresh();
+      this._dirty = false;
+    }
+    this._active = true;
+  };
+
+  /**
+   * Activate. invoked when visualization is de selected
+   */
+  deactivate() {
+    this._active = false;
+  };
+
+  /**
+   * Is active
+   */
+  isActive() {
+    return this._active;
+  };
+
+  /**
+   * When window or paragraph is resized
+   */
+  resize() {
+    if (this.isActive()) {
+      this.refresh();
+    } else {
+      this._dirty = true;
+    }
+  };
+
+  /**
+   * Set new config
+   */
+  setConfig(config) {
+    this.config = config;
+    if (this.isActive()) {
+      this.refresh();
+    } else {
+      this._dirty = true;
+    }
+  };
+
+  /**
+   * Emit config. config will sent to server and saved.
+   */
+  emitConfig(config) {
+    this._emitter(config);
+  };
+
+  /**
+   * render setting
+   */
+  renderSetting(targetEl) {
+    var setting = this.getSetting();
+    if (!setting) {
+      return;
+    }
+
+    // already readered
+    if (this._scope) {
+      var self = this;
+      this._scope.$apply(function() {
+        for (var k in setting.scope) {
           self._scope[k] = setting.scope[k];
         }
-      }
-    });
-    return;
-  } else {
-    this._prevSettingScope = setting.scope;
-  }
-
-  var scope = this._createNewScope();
-  for (var k in setting.scope) {
-    scope[k] = setting.scope[k];
-  }
-  var template = setting.template;
-
-  if (template.split('\n').length === 1 &&
-      template.endsWith('.html')) { // template is url
-    var self = this;
-    this._templateRequest(template).then(function(t) {
-      self._renderSetting(targetEl, t, scope);
-    });
-  } else {
-    this._renderSetting(targetEl, template, scope);
-  }
-};
-
-zeppelin.Visualization.prototype._renderSetting = function(targetEl, template, scope) {
-  this._targetEl = targetEl;
-  targetEl.html(template);
-  this._compile(targetEl.contents())(scope);
-  this._scope = scope;
-};
+
+        for (var k in self._prevSettingScope) {
+          if (!setting.scope[k]) {
+            self._scope[k] = setting.scope[k];
+          }
+        }
+      });
+      return;
+    } else {
+      this._prevSettingScope = setting.scope;
+    }
+
+    var scope = this._createNewScope();
+    for (var k in setting.scope) {
+      scope[k] = setting.scope[k];
+    }
+    var template = setting.template;
+
+    if (template.split('\n').length === 1 &&
+        template.endsWith('.html')) { // template is url
+      var self = this;
+      this._templateRequest(template).then(function(t) {
```

Comment:
instead of confusing `self = this` it is better to use `=>`, e.g. `templateRequest().then(t => {/* use this here safely */})`; it will bind the context of the outer scope